[バグ] Shelem: 落札者の捨て札のカード点が消える（ラウンド合計が100点にならず、配りによってテストが落ちる）

## 背景・現状

`TestShelem_RoundPointsAlwaysSumToOneHundred` が **配りによって落ちる**。
PR #5794 の CI で観測 ([run 31937944594](https://github.com/yuta-yoshinaga/go_trumpcards/actions/runs/31937944594)):

```
--- FAIL: TestShelem_RoundPointsAlwaysSumToOneHundred (0.00s)
    Shelem_test.go:379:
        	expected: 100
        	actual  : 95
```

当該 PR は `internal/domain` を 1 行も触っていない (`git diff develop...HEAD -- internal/domain/` が空)。
**フレークではなく、develop に既存の実装バグ**である。

## 原因

カード点は 52 枚全体に散っている (`ShelemCardPoints`):

| ランク | 点 | 枚数 | 小計 |
|---|---:|---:|---:|
| A | 10 | 4 | 40 |
| 10 | 10 | 4 | 40 |
| 5 | 5 | 4 | 20 |
| | | | **100** |

配りは 4人 × 12枚 = 48枚 + **ウィドウ 4枚** = 52枚。落札者はウィドウを取り込み
**4 枚捨てる**。ところが捨て札は `internal/domain/Shelem.go` の

```go
for _, i := range sorted {
    p.RemoveCard(i)
}
```

で**破棄されるだけ**で、どのチームにも加算されない。`roundPoints` が増えるのは
`resolveTrick()` の 1 箇所 (`Shelem.go:592`) しかなく、捨て札はそこを通らない。

したがって**点札が捨て札に混ざった瞬間、そのラウンドのカード点合計は 100 未満になる**。
今回は 5 が 1 枚捨てられて 95。A か 10 が捨てられれば 90 になる。

**なぜ「たまに」しか落ちないか**: CPU の捨て札ヒューリスティックが点札を避けるため
(`score := shelemRank(c) + ShelemCardPoints(c)*10`)、通常は 0 点札から捨てる。
落札者の 16 枚が点札に偏り、0 点の捨て札を 4 枚作れない配りのときだけ表面化する。

## 提案

ウィドウ・キティ系のゲームの通例どおり、**捨て札のカード点を落札者チームに加算する**。
`PlayerDiscard` と `cpuDiscardAndTrump` の両方の経路で、`RemoveCard` する前に
`ShelemCardPoints` を合計して `s.roundPoints[ShelemTeamOf(s.declarerIdx)]` に足す。

（別解として「点札を捨てられないようにする」も考えられるが、落札者の手が点札だらけの配りでは
4 枚の 0 点札を作れないことがあるため、規則として成立しない。）

## 受け入れ条件

- [ ] 捨て札に点札が含まれても、`GetRoundPoints(0) + GetRoundPoints(1) == 100` が常に成立する
- [ ] 人間の `PlayerDiscard` と CPU の `cpuDiscardAndTrump` の**両方**の経路で加算される
- [ ] 加算先は落札者のチーム
- [ ] **点札を強制的に捨てさせる決定的なテスト**を追加する（配り任せにしない。
      現状のテストは配り依存なので、修正が入っても偶然通るだけでは検証にならない）
- [ ] 既存の `TestShelem_RoundPointsAlwaysSumToOneHundred` は残す（回帰の網として有効）
- [ ] 契約達成判定 (`Shelem.go:655` 付近の `got := s.roundPoints[declTeam]`) が
      加算後の値で正しく動くことを確認する — 落札者の獲得点が増えるため、
      **契約成立の判定が変わる**。ここはルール上の意図を確認してから決めること

## 対象ファイル

- `internal/domain/Shelem.go`
- `internal/domain/Shelem_test.go`

---
_PR #5794 の CI で観測。flake-ledger に記録したうえで「1 件目は実バグとして調べる」規則に従って調査した結果、実バグと判明したもの。_

--- a/internal/domain/Shelem.go
+++ b/internal/domain/Shelem.go
@@ -419,17 +419,38 @@ func (s *Shelem) PlayerDiscard(indices []int, suit int) error {
 			}
 		}
 	}
+	discarded := make([]*Card, 0, len(sorted))
 	for _, i := range sorted {
+		discarded = append(discarded, p.GetCard(i))
 		p.RemoveCard(i)
 	}
+	s.creditDiscard(discarded)
 	s.acceptTrump(suit)
 	return nil
+}
+
+// creditDiscard は捨て札のカード点を落札者チームに加える。
+//
+// カード点は 52 枚全体で ShelemHandPoints (100) 点あり、そのうち 4 枚はウィドウに
+// 伏せられている。落札者がウィドウを取り込んで 4 枚捨てたあと、その札の点を誰にも
+// 加算しないと、点札が捨て札に入った分だけラウンド合計が 100 を割る (#5795)。
+// ウィドウ・キティ系の通例どおり、捨てた点は落札者チームのものとして扱う。
+func (s *Shelem) creditDiscard(cards []*Card) {
+	if s.declarerIdx < 0 {
+		return
+	}
+	pts := 0
+	for _, c := range cards {
+		pts += ShelemCardPoints(c)
+	}
+	s.roundPoints[ShelemTeamOf(s.declarerIdx)] += pts
 }
 
 // cpuDiscardAndTrump CPU の落札者が点にならない札を捨て、長いスートを切り札にする
 func (s *Shelem) cpuDiscardAndTrump() {
 	suit := s.longestSuit(s.declarerIdx)
 	p := s.players[s.declarerIdx]
+	discarded := make([]*Card, 0, ShelemWidowSize)
 	// **切り札でなく、点にもならない札から捨てる。**
 	for p.GetCardsSize() > ShelemHandSize {
 		worst, worstScore := 0, 1<<30
@@ -443,8 +464,10 @@ func (s *Shelem) cpuDiscardAndTrump() {
 				worst, worstScore = i, score
 			}
 		}
+		discarded = append(discarded, p.GetCard(worst))
 		p.RemoveCard(worst)
 	}
+	s.creditDiscard(discarded)
 	s.acceptTrump(suit)
 }
 

--- a/internal/domain/Shelem.go
+++ b/internal/domain/Shelem.go
@@ -436,9 +436,9 @@ func (s *Shelem) PlayerDiscard(indices []int, suit int) error {
 // 加算しないと、点札が捨て札に入った分だけラウンド合計が 100 を割る (#5795)。
 // ウィドウ・キティ系の通例どおり、捨てた点は落札者チームのものとして扱う。
 func (s *Shelem) creditDiscard(cards []*Card) {
-	if s.declarerIdx < 0 {
-		return
-	}
+	// declarerIdx の防御は置かない。PlayerDiscard は declarerIdx != 0 で早期 return し、
+	// cpuDiscardAndTrump は declarerIdx != 0 の分岐の中でしか呼ばれない。到達しない
+	// 分岐は codecov にも残る。
 	pts := 0
 	for _, c := range cards {
 		pts += ShelemCardPoints(c)

--- a/internal/domain/Shelem_test.go
+++ b/internal/domain/Shelem_test.go
@@ -708,3 +708,127 @@ func TestShelem_ActionLog(t *testing.T) {
 	}
 	assert.True(t, kinds["bid"])
 }
+
+// 捨て札に混ざったカード点が消えないこと。
+//
+// カード点は 52 枚全体で 100 点 (A/10 が 10 点 ×8 枚、5 が 5 点 ×4 枚)。配りは
+// 48 枚 + ウィドウ 4 枚で、落札者はウィドウを取り込んで 4 枚捨てる。捨て札の点を
+// 誰にも加算しないと、点札が捨て札に入った瞬間ラウンド合計が 100 を割る (#5795)。
+//
+// TestShelem_RoundPointsAlwaysSumToOneHundred は配り依存でしか露出しないので、
+// ここは点札を強制的に捨てさせて決定的に確かめる。
+func TestShelem_DiscardedCardPointsGoToTheDeclarer(t *testing.T) {
+	newDiscardShelem := func(t *testing.T) *Shelem {
+		t.Helper()
+		s := newTestShelem(t)
+		s.SetContractForTest(0, ShelemMinBid, false)
+		s.SetPhaseForTest(ShelemPhaseDiscard)
+		return s
+	}
+
+	t.Run("human discard credits the declarer's team", func(t *testing.T) {
+		s := newDiscardShelem(t)
+		// 先頭 4 枚を点札にして、それを捨てさせる: 10 + 10 + 5 + 5 = 30。
+		hand := []*Card{
+			NewCard(CardDesignSpade, 1, false),   // A = 10
+			NewCard(CardDesignHeart, 10, false),  // 10 = 10
+			NewCard(CardDesignClover, 5, false),  // 5  = 5
+			NewCard(CardDesignDiamond, 5, false), // 5  = 5
+		}
+		for i := range ShelemHandSize {
+			hand = append(hand, NewCard(CardDesignSpade, (i%4)+6, false)) // 6-9 のみ = 確実に 0 点
+		}
+		shelemHandOf(s, 0, hand...)
+
+		require.NoError(t, s.PlayerDiscard([]int{0, 1, 2, 3}, CardDesignSpade))
+
+		assert.Equal(t, 30, s.GetRoundPoints(ShelemTeamOf(0)),
+			"捨て札のカード点が落札者チームに入っていない")
+		assert.Equal(t, 0, s.GetRoundPoints(1-ShelemTeamOf(0)),
+			"相手チームには入らない")
+	})
+
+	t.Run("discarding only zero-point cards adds nothing", func(t *testing.T) {
+		s := newDiscardShelem(t)
+		hand := make([]*Card, 0, ShelemHandSize+ShelemWidowSize)
+		for i := range ShelemHandSize + ShelemWidowSize {
+			// 6-9 のみ。(i%7)+2 は 5 を含み、5 は 5 点なので 0 点札ではない。
+			hand = append(hand, NewCard(CardDesignSpade, (i%4)+6, false))
+		}
+		shelemHandOf(s, 0, hand...)
+
+		require.NoError(t, s.PlayerDiscard([]int{0, 1, 2, 3}, CardDesignSpade))
+
+		assert.Equal(t, 0, s.GetRoundPoints(ShelemTeamOf(0)))
+	})
+
+	t.Run("cpu discard credits the declarer's team too", func(t *testing.T) {
+		s := newTestShelem(t)
+		s.SetContractForTest(1, ShelemMinBid, false)
+		s.SetPhaseForTest(ShelemPhaseDiscard)
+		// CPU は点札を避けて捨てるので、点札しか持たせない配りを組む。
+		hand := make([]*Card, 0, ShelemHandSize+ShelemWidowSize)
+		for range ShelemHandSize + ShelemWidowSize {
+			hand = append(hand, NewCard(CardDesignHeart, 5, false)) // すべて 5 点
+		}
+		shelemHandOf(s, 1, hand...)
+
+		s.CpuDiscardAndTrumpForTest()
+
+		assert.Equal(t, ShelemWidowSize*5, s.GetRoundPoints(ShelemTeamOf(1)),
+			"CPU 経路でも捨て札の点が落札者チームに入る")
+	})
+
+	t.Run("the round still totals 100 when points are discarded", func(t *testing.T) {
+		s := newDiscardShelem(t)
+		hand := []*Card{
+			NewCard(CardDesignSpade, 1, false),
+			NewCard(CardDesignHeart, 10, false),
+			NewCard(CardDesignClover, 5, false),
+			NewCard(CardDesignDiamond, 5, false),
+		}
+		for i := range ShelemHandSize {
+			hand = append(hand, NewCard(CardDesignSpade, (i%4)+6, false))
+		}
+		shelemHandOf(s, 0, hand...)
+		require.NoError(t, s.PlayerDiscard([]int{0, 1, 2, 3}, CardDesignSpade))
+
+		// 残る 70 点が卓上で取り合われたとみなす。
+		s.SetRoundPointsForTest(1-ShelemTeamOf(0), 70)
+		assert.Equal(t, ShelemHandPoints,
+			s.GetRoundPoints(0)+s.GetRoundPoints(1),
+			"捨て札の点を含めてラウンド合計が 100 になる")
+	})
+}
+
+// 捨て札の点は契約達成の判定にも数える。
+//
+// finishRound は `got := s.roundPoints[declTeam]` を契約と比べるので、捨て札を
+// 加算した分だけ落札者は契約に近づく。ウィドウ・キティ系の通例どおりで意図的な
+// 挙動だが、#5795 の修正で判定が変わる箇所なのでテストで固定しておく。
+func TestShelem_DiscardedPointsCountTowardTheContract(t *testing.T) {
+	s := newTestShelem(t)
+	s.SetContractForTest(0, 100, false) // 卓上だけでは届かない契約にする
+	s.SetPhaseForTest(ShelemPhaseDiscard)
+
+	hand := []*Card{
+		NewCard(CardDesignSpade, 1, false),   // 10
+		NewCard(CardDesignHeart, 1, false),   // 10
+		NewCard(CardDesignClover, 1, false),  // 10
+		NewCard(CardDesignDiamond, 1, false), // 10
+	}
+	for i := range ShelemHandSize {
+		hand = append(hand, NewCard(CardDesignSpade, (i%4)+6, false))
+	}
+	shelemHandOf(s, 0, hand...)
+	require.NoError(t, s.PlayerDiscard([]int{0, 1, 2, 3}, CardDesignSpade))
+	require.Equal(t, 40, s.GetRoundPoints(ShelemTeamOf(0)), "捨て札 40 点が入っている")
+
+	// 卓上で 60 点取れば、捨て札の 40 と合わせてちょうど契約 100 に届く。
+	s.SetRoundPointsForTest(ShelemTeamOf(0), 100)
+	s.SetTrickNumberForTest(ShelemTricksPerRound)
+	s.FinishRoundForTest()
+
+	assert.Positive(t, s.GetScore(ShelemTeamOf(0)),
+		"契約達成として加点される（捨て札の点を数えないと未達になる）")
+}

--- a/internal/domain/Shelem_testhelpers.go
+++ b/internal/domain/Shelem_testhelpers.go
@@ -71,3 +71,6 @@ func (s *Shelem) PlayForTest(playerIdx, cardIndex int) error { return s.play(pla
 
 // CpuChoiceForTest CPU が選ぶ手札インデックスを返す
 func (s *Shelem) CpuChoiceForTest(playerIdx int) int { return s.chooseCpuCard(playerIdx) }
+
+// CpuDiscardAndTrumpForTest CPU 落札者の捨て札・切り札決定を直接呼ぶ
+func (s *Shelem) CpuDiscardAndTrumpForTest() { s.cpuDiscardAndTrump() }

